rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -682,35 +712,26 @@ bool PlayerImpl::play_next()
   // Use RCLCPP_DEBUG_STREAM to avoid delays in the burst mode
   RCLCPP_DEBUG_STREAM(owner_->get_logger(), "Playing next message.");
 
-  // Temporary take over playback from play_messages_from_queue()
-  std::lock_guard<std::mutex> main_play_loop_lk(skip_message_in_main_play_loop_mutex_);
-  // Check one more time that we are in pause mode after waiting on mutex. Someone could call
-  // resume() or stop() from another thread while we were waiting on mutex.
-  if (!clock_->is_paused()) {
-    RCLCPP_WARN_STREAM(owner_->get_logger(), "Called play next, but not in paused state.");
-    return false;
-  }
-  skip_message_in_main_play_loop_ = true;
   // Wait for player to be ready for playback messages from queue i.e. wait for Player:play() to
   // be called if not yet and queue to be filled with messages.
   {
     std::unique_lock<std::mutex> lk(ready_to_play_from_queue_mutex_);
     ready_to_play_from_queue_cv_.wait(lk, [this] {return is_ready_to_play_from_queue_;});
   }
 
-  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = peek_next_message_from_queue();
-
-  bool next_message_published = false;
-  while (rclcpp::ok() && !next_message_published && !stop_playback_ &&
-    message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
-  {
-    next_message_published = publish_message(message_ptr);
-    clock_->jump(message_ptr->recv_timestamp);
+  // Request to play next
+  play_next_ = true;
 
-    message_queue_.pop();
-    message_ptr = peek_next_message_from_queue();
-  }
-  return next_message_published;
+  // Wait for play next to be done, and then return the result
+  std::unique_lock<std::mutex> lk(finished_play_next_mutex_);
+  // Temporarily resume clock to force wakeup in clock_->sleep_until(time),
+  // then return in pause mode to preserve original state of the player
+  clock_->resume();
+  clock_->pause();
```

Comment:
I will check if we can wake up `clock_->sleep_until()` another way, maybe through a new clock API.

Reply:
Looks like just calling `std::condition_variable::notify_all()` on the clock's condition variable should wake it up from its sleep (i.e., [`std::condition_variable::wait_for()`](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for) and [`std::condition_variable::wait_until()`](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until)):

> The thread will be unblocked when [notify_all()](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all) or [notify_one()](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) is executed, or abs_time is reached. It may also be unblocked spuriously.

Reply:
I opened a separate PR for this: #1869.